x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotRetentionTask.java

```diff
@@ -0,0 +1,99 @@
+
+/**
+ * The {@code SnapshotRetentionTask} is invoked by the scheduled job from the
+ * {@link SnapshotRetentionService}. It is responsible for retrieving the snapshots for repositories
+ * that have an SLM policy configured, and then deleting the snapshots that fall outside the
+ * retention policy.
+ */
+public class SnapshotRetentionTask implements SchedulerEngine.Listener {
+
+    private static final Logger logger = LogManager.getLogger(SnapshotRetentionTask.class);
+    private static final AtomicBoolean running = new AtomicBoolean(false);
+
+    private final Client client;
+    private final ClusterService clusterService;
+
+    public SnapshotRetentionTask(Client client, ClusterService clusterService) {
+        this.client = client;
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        assert event.getJobName().equals(SnapshotRetentionService.SLM_RETENTION_JOB_ID) :
+            "expected id to be " + SnapshotRetentionService.SLM_RETENTION_JOB_ID + " but it was " + event.getJobName();
+        if (running.compareAndSet(false, true)) {
+            try {
+                logger.info("starting SLM retention snapshot cleanup task");
+                final ClusterState state = clusterService.state();
+
+                // Find all SLM policies that have retention enabled
+                final Map<String, SnapshotLifecyclePolicy> policiesWithRetention = getAllPoliciesWithRetentionEnabled(state);
+
+                // For those policies (there may be more than one for the same repo),
+                // return the repos that we need to get the snapshots for
+                final Set<String> repositioriesToFetch = policiesWithRetention.values().stream()
+                    .map(SnapshotLifecyclePolicy::getRepository)
+                    .collect(Collectors.toSet());
+
+                // Find all the snapshots that are past their retention date
+                // TODO: include min/max snapshot count as a criteria for deletion also
+                final List<SnapshotInfo> snapshotsToBeDeleted = getAllSnapshots(repositioriesToFetch).stream()
+                    .filter(snapshot -> snapshotEligibleForDeletion(snapshot, policiesWithRetention))
+                    .collect(Collectors.toList());
+
+                // Finally, delete the snapshots that need to be deleted
+                deleteSnapshots(snapshotsToBeDeleted);
+
+            } finally {
+                running.set(false);
+            }
+        } else {
+            logger.debug("snapshot lifecycle retention task started, but a task is already running, skipping");
+        }
+    }
+
+    static Map<String, SnapshotLifecyclePolicy> getAllPoliciesWithRetentionEnabled(final ClusterState state) {
+        // TODO: fill me in
+        return Collections.emptyMap();
+    }
+
+    static boolean snapshotEligibleForDeletion(SnapshotInfo snapshot, Map<String, SnapshotLifecyclePolicy> policies) {
```

Comment:
It's not super relevant given that a later PR can refactor this as necessary but I don't think this method signature is going to work given that the plan is to have a min/max snapshot count (vs. just considering snapshot age).

Reply:
I agree, I figured this would grow/change as it was actually implemented